Core API: findNodes for composite and string queries

fickludd opened the pull request:
This PR bring Core API up to date with underlying index capabilities. The added methods only rely on functionality which is anyway required by Cypher, so it should not constrain future product development. 

changelog: Extended Core API with more variants of findNodes: String prefix, suffix and contains queries on single property indexes, and exact seeks in composite indexes.

fickludd committed:
Prefix index search via Core API

fickludd committed:
Fix bug in index suffix and contains queries with in-tx modified nodes

fickludd committed:
Implement index EXACT, SUFFIX and CONTAINS queries over Core API

fickludd committed:
Test String findNodes with and without a backing index.

fickludd committed:
Implemented composite findNodes in Core API

fickludd committed:
checkstyle

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -650,11 +719,103 @@ private InternalTransaction beginTransactionInternal( KernelTransaction.Type typ
+            return emptyResourceIterator();
+        }
+
+        int[] propertyIds = getSortedPropertyIds( queries );
```
I just realized, before doing the whole sorted properties dance, we should make a quick check to see if there is an index the the properties in the order they were provided. That would likely match in many real situations, and should be much faster. 

If no index matches for the given query order, we can fall back to the sorted dance.

fickludd committed:
Provide default implementation of new methods for backwards compatibility

craigtaverner commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java:
```diff
@@ -134,6 +134,133 @@
+     * scan all labeled nodes looking for matching nodes.
+     * <p>
+     * Note that equality for values do not follow the rules of Java. This means that the number 42 is equals to all
+     * other 42 numbers, indifferently of if they are encoded as Integer, Long, Float, Short, Byte or Double.
```
The phase _indifferently of if_ should be _regardless of whether_. I see, however, that this was copied from elsewhere, so that would be a fix to much older javadocs.

craigtaverner commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java:
```diff
@@ -134,6 +134,133 @@
+     * The search mode and value template are used to select nodes of interest. The search mode can
+     * be one of
+     * <ul>
+     *   <li>EXACT: The value has to match the template exactly.</li>
```
Perhaps worth mentioning that EXACT is the same as the behaviour of the older findNodes method?

craigtaverner commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/StringSearchMode.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    SUFFIX,
+    /**
+     * The value must contain the template. Only exact matches are supported.
```
What does 'only exact matches are supported mean'? Sounds like it means CONTAINS==EXACT, which makes CONTAINS redunant. I assume this is not what it means. Either clarify or remove that sentence.

MishaDemianenko commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java:
```diff
@@ -134,6 +134,133 @@
+    /**
+     * Returns all nodes having the label, and the wanted property values.
+     * If an online index is found, it will be used to look up the requested
+     * nodes. The specified properties do not have to be provided in the same order
```
do not think we need to mention order here

MishaDemianenko commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java:
```diff
@@ -134,6 +134,133 @@
+     */
+    default ResourceIterator<Node> findNodes( Label label, String key1, Object value1, String key2, Object value2 )
+    {
+        throw new UnsupportedOperationException( "Composite findNodes is not supported by this GraphDatabaseService" );
```
I think would be good to rephrase this kind of messages to be something like "findNodes by multiple property names and values are not supported."

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/TxState.java:
```diff
@@ -1014,6 +1015,33 @@ public PrimitiveLongReadableDiffSets indexUpdatesForScan( SchemaIndexDescriptor
+    @Override
+    public PrimitiveLongReadableDiffSets indexUpdatesForSuffixOrContains( SchemaIndexDescriptor descriptor, IndexQuery query )
+    {
+        descriptor.schema().getPropertyId(); // assert single property index
```
please do not perform validation based on method call side effects, if you need to validate number of properties just do that

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -650,11 +719,103 @@ private InternalTransaction beginTransactionInternal( KernelTransaction.Type typ
+        Statement statement = transaction.acquireStatement();
+        Read read = transaction.dataRead();
+
+        if ( isInvalidQuery( labelId, queries ) )
```
there is an assumption that its impossible to have two queries for the same property key, would be good to check re-validate that

fickludd replied:
Added a check for duplicates.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -650,11 +719,103 @@ private InternalTransaction beginTransactionInternal( KernelTransaction.Type typ
+        int[] propertyIds = getSortedPropertyIds( queries );
+        int[] workingCopy = new int[propertyIds.length];
+
+        Iterator<CapableIndexReference> indexes = transaction.schemaRead().indexesGetForLabel( labelId );
```
we have a method 'storeReadLayer.indexGetForSchema' that know how to find index by label and all property keys, maybe using that instead will be actually faster than looking for each matching index separately?

fickludd replied:
Yes, the `storeaReadLayer.indexGetForSchema` is reachable by `SchemaRead.index`, which I now added as a first attempt at a fast path. However this method does not try to reorder properties because no-one else needs that, so we still need to retain the `indexesGetForLabel` to try to find out-of-order indexes.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -862,17 +1019,26 @@ void closeResources( Statement statement )
         }
 
-        private boolean hasPropertyWithValue()
+        private boolean hasPropertiesWithValues()
```
think this looks simpler:
```
        private boolean hasPropertiesWithValues()
        {
            read.singleNode( nodeLabelCursor.nodeReference(), nodeCursor );
            if ( nodeCursor.next() )
            {
                nodeCursor.properties( propertyCursor );
                while ( propertyCursor.next() )
                {
                    for ( IndexQuery query : queries )
                    {
                        if ( propertyCursor.propertyKey() != query.propertyKeyId() || !query.acceptsValueAt( propertyCursor ) )
                        {
                            return false;
                        }
                    }
                }
                return true;
            }
            return false;
        }
```

fickludd replied:
I don't think this is doing the same thing? The node may have properties which are not queried for, as long as it also has every queried for property. Here we would `return false` on the first non-queried for property, which would exclude nodes incorrectly.

However, you made me think more about this, and actually if we find a propertyKey matching a query, _but with the wrong value_, we can return early. Added that.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/DefaultNodeValueIndexCursor.java:
```diff
@@ -75,19 +75,23 @@ public void initialize( SchemaIndexDescriptor descriptor, IndexProgressor progre
 
         case stringPrefix:
-            prefixQuery( descriptor, (IndexQuery.StringPrefixPredicate) query[0] );
+            prefixQuery( descriptor, (IndexQuery.StringPrefixPredicate) firstPredicate );
```
looks like  assert query.length == 1; is missing here

fickludd replied:
yupp :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/TxState.java:
```diff
@@ -1014,6 +1015,33 @@ public PrimitiveLongReadableDiffSets indexUpdatesForScan( SchemaIndexDescriptor
+        PrimitiveLongDiffSets diffs = new PrimitiveLongDiffSets();
+        for ( Map.Entry<ValueTuple,PrimitiveLongDiffSets> entry : updates.entrySet() )
+        {
+            if ( query.acceptsValue( entry.getKey().getOnlyValue() ) )
```
why is this check only part of this retrieval of index updates?

fickludd replied:
For `seek`, `prefix`, `range` and `scan`, these methods can query the `updates` map (a `HashMap` or `TreeMap`) for updates which exactly match the query - meaning that no filtering is needed.

For `suffix` and `contains` that is not possible though, so we need to scan all updates and filter for keys which actually match the query. This filtering was not done previously, and that bug is what we fixed.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/graphdb/IndexingCompositeQueryAcceptanceTest.java:
```diff
@@ -0,0 +1,316 @@
+        db = dbRule.getGraphDatabaseAPI();
+        if ( withIndex )
+        {
+            Neo4jMatchers.createIndex( db, LABEL, keys );
```
why on <strike>Earth</strike> any planet! we will gonna have a method on Neo4jMatchers! that does index creation!

fickludd replied:
Yes... that was surprising to me too. Changed it to use Core API in the new test classes.

craigtaverner replied:
I traced this to a commit by Ian in 2013. Seems to be part of a big push in expanding the testing framework. And it is used all over the place. I think cleaning this up is an entirely separate job, not for this PR.

MishaDemianenko commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/StringSearchMode.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    SUFFIX,
+    /**
+     * The value must contain the template. Only exact matches are supported.
```
not sure what "Only exact matches are supported" means here?

fickludd replied:
Reworded it. I was trying to convey that regular expressions are not supported in case someone tried that.

craigtaverner committed:
Also delete schema in database cleaning code

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/api/schema/LabelSchemaDescriptor.java:
```diff
@@ -88,7 +88,7 @@ public int getPropertyId()
         {
             throw new IllegalStateException(
-                    "Single property schema requires one property but had " + propertyIds.length );
+                    "Single property schema required one property but had " + propertyIds.length );
```
The previous version was more correct. The requirement is still in place, so present tense should be used. The new version implies that the schema _used to require_, a requirement that is no longer in force, which is not true.

fickludd committed:
Improvements from review

 - Misc javadoc comment clarifications
 - some refactoring of composite findNodes

fickludd committed:
Merge branch '3.4-find-more-nodes' of github.com:fickludd/neo4j into 3.4-find-more-nodes

fickludd committed:
javadoc improvements

fickludd committed:
assert the correct thing
